src/main/java/se/liu/ida/hefquin/queryplan/executable/impl/ops/ExecOpGenericJoin.java

```diff
@@ -0,0 +1,58 @@
+package se.liu.ida.hefquin.queryplan.executable.impl.ops;
+
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.data.SolutionMapping;
+import se.liu.ida.hefquin.data.jenaimpl.JenaBasedSolutionMapping;
+import se.liu.ida.hefquin.queryplan.executable.IntermediateResultBlock;
+import se.liu.ida.hefquin.queryplan.executable.IntermediateResultElementSink;
+import org.apache.jena.sparql.engine.join.JoinKey;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class ExecOpGenericJoin implements BinaryExecutableOp {
+    protected List<Var> keys;
```

Comment:
This member variable seems weird.

First of all, it does not seem to be initialized properly.

Second, should it really be mutable? Are there any guarantees about when does it change, and through which means does it change? 

Third, the name does not really make clear what this actually is about. What are "keys" in the context of join algorithms? I assume you mean "join variables"?